[Nintendo Switch] Crash on start

### Installation Source

Self-compiled from source

### PvZ-Portable Version

0.1.16

### Platform

Nintendo Switch

### OS / Environment Info

Nintendo Switch

### Build Dependencies (self-compiled only)

_No response_

### Resource Pack Version

1.2.0.1073 (Origin / PopCap GOTY)

### Bug Description

Atmosphère Crash Report (v1.7):
Result:                          0x2A8 (2168-0001)

Process Info:
    Process Name:                Application
    Program ID:                  03db12780bd84000
    Process ID:                  00000000000000d0
    Process Flags:               00000073
    User Exception Address:      0000000028095000
Exception Info:
    Type:                        Instruction Abort
    Address:                     0000000000000000
Crashed Thread Info:
    Thread ID:                   0000000000000b35
    Stack Region:                000000004a348000-000000004a448000
    Registers:
        X[00]:                   0000000000008b31
        X[01]:                   0000000f3c6581a0
        X[02]:                   0000000000000000
        X[03]:                   0000000000000fff
        X[04]:                   0000000000000000
        X[05]:                   0000000000000000
        X[06]:                   6d68656463220964
        X[07]:                   7f7f7f7f7f7f7f7f
        X[08]:                   0101010101010101
        X[09]:                   696e69206e656562
        X[10]:                   206d657473797362
        X[11]:                   20746f6e20736168
        X[12]:                   696e69206e656562
        X[13]:                   64657a696c616974
        X[14]:                   787a79727d6e1f19
        X[15]:                   0000000000000000
        X[16]:                   0000000f3c003bc0 (pvz-portable + 0x22fbc0)
        X[17]:                   0000000000000000
        X[18]:                   000000004a447a38
        X[19]:                   0000000f3c658180
        X[20]:                   0000000000000209
        X[21]:                   0000000f3c910000
        X[22]:                   0000000f3bf28000 (pvz-portable + 0x154000)
        X[23]:                   000000004a447ef0
        X[24]:                   0000000000000000
        X[25]:                   0000000043e09010
        X[26]:                   ffffffffffffffff
        X[27]:                   0000000043e010c0
        X[28]:                   000000004a448000
        FP:                      000000004a447cf0
        LR:                      0000000f3bf372ac (pvz-portable + 0x1632ac)
        SP:                      000000004a447cf0
        PC:                      0000000000000000
    Stack Trace:
        ReturnAddress[00]:       0000000f3bf3f004 (pvz-portable + 0x16b004)
        ReturnAddress[01]:       0000000f3bf2d674 (pvz-portable + 0x159674)
        ReturnAddress[02]:       0000000f3bf8a344 (pvz-portable + 0x1b6344)
        ReturnAddress[03]:       0000000f3bf34618 (pvz-portable + 0x160618)
        ReturnAddress[04]:       0000000f3be92e0c (pvz-portable + 0xbee0c)
        ReturnAddress[05]:       0000000f3bdd8860 (pvz-portable + 0x4860)
        ReturnAddress[06]:       0000000f3bdd7700 (pvz-portable + 0x3700)
        ReturnAddress[07]:       0000000000000000
    Stack Dump:                               00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f
                                 00004a447cf0 60 7d 44 4a 00 00 00 00 04 f0 f3 3b 0f 00 00 00
                                 00004a447d00 e0 06 ff 6d 0c 00 00 00 e0 06 ff 6d 0c 00 00 00
                                 00004a447d10 00 80 f2 3b 0f 00 00 00 00 80 f2 3b 0f 00 00 00
                                 00004a447d20 10 90 e0 43 00 00 00 00 ff ff ff ff ff ff ff ff
                                 00004a447d30 26 00 00 00 3d 00 00 00 09 02 00 00 0c 00 00 00
                                 00004a447d40 58 81 65 3c 0f 00 00 00 80 81 65 3c 0f 00 00 00
                                 00004a447d50 20 82 65 3c 0f 00 00 00 00 00 00 00 00 00 00 00
                                 00004a447d60 00 7e 44 4a 00 00 00 00 74 d6 f2 3b 0f 00 00 00
                                 00004a447d70 e0 7b a1 65 0c 00 00 00 e0 06 ff 6d 0c 00 00 00
                                 00004a447d80 00 80 f2 3b 0f 00 00 00 00 80 f2 3b 0f 00 00 00
                                 00004a447d90 f0 7e 44 4a 00 00 00 00 00 00 00 00 00 00 00 00
                                 00004a447da0 10 90 e0 43 00 00 00 00 ff ff ff ff ff ff ff ff
                                 00004a447db0 f0 7e 44 4a 00 00 00 00 00 00 00 00 00 00 00 00
                                 00004a447dc0 10 90 e0 43 00 00 00 00 ff ff ff ff ff ff ff ff
                                 00004a447dd0 c0 7e 44 4a 00 00 00 00 00 80 f2 3b 0f 00 00 00
                                 00004a447de0 10 01 00 00 00 00 00 00 70 90 ab 6e 0c 00 00 00
    TLS Address:                 0000000028095200
    TLS Dump:                                 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f
                                 000028095200 00 00 01 00 09 08 e0 03 00 00 28 00 f0 79 44 4a
                                 000028095210 53 46 43 4f 00 00 00 00 00 00 00 00 99 5d e2 a4
                                 000028095220 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095230 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00
                                 000028095240 03 00 75 01 06 41 28 c0 00 00 00 00 00 00 00 00
                                 000028095250 28 00 75 01 f0 79 44 4a 00 00 28 00 00 00 00 00
                                 000028095260 00 00 00 00 18 00 00 00 58 7b 44 4a 00 00 18 00
                                 000028095270 f0 00 00 00 00 00 00 00 d0 00 00 00 00 00 00 00
                                 000028095280 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095290 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952a0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952b0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952c0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952d0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952e0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952f0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Module Info:
    Number of Modules:           01
    Module 00:
        Address:                 0000000f3bdd4000-0000000f3c643000
        Name:                    pvz-portable
        Module Id:               3A85F65B2D4129D7F1DA888BC692A350BF9B4874000000000000000000000000
Thread Report:
Number of Threads:               02
Threads[00]:
    Thread ID:                   0000000000000b35
    Stack Region:                000000004a348000-000000004a448000
    Registers:
        X[00]:                   0000000000008b31
        X[01]:                   0000000f3c6581a0
        X[02]:                   0000000000000000
        X[03]:                   0000000000000fff
        X[04]:                   0000000000000000
        X[05]:                   0000000000000000
        X[06]:                   6d68656463220964
        X[07]:                   7f7f7f7f7f7f7f7f
        X[08]:                   0101010101010101
        X[09]:                   696e69206e656562
        X[10]:                   206d657473797362
        X[11]:                   20746f6e20736168
        X[12]:                   696e69206e656562
        X[13]:                   64657a696c616974
        X[14]:                   787a79727d6e1f19
        X[15]:                   0000000000000000
        X[16]:                   0000000f3c003bc0 (pvz-portable + 0x22fbc0)
        X[17]:                   0000000000000000
        X[18]:                   000000004a447a38
        X[19]:                   0000000f3c658180
        X[20]:                   0000000000000209
        X[21]:                   0000000f3c910000
        X[22]:                   0000000f3bf28000 (pvz-portable + 0x154000)
        X[23]:                   000000004a447ef0
        X[24]:                   0000000000000000
        X[25]:                   0000000043e09010
        X[26]:                   ffffffffffffffff
        X[27]:                   0000000043e010c0
        X[28]:                   000000004a448000
        FP:                      000000004a447cf0
        LR:                      0000000f3bf372ac (pvz-portable + 0x1632ac)
        SP:                      000000004a447cf0
        PC:                      0000000000000000
    Stack Trace:
        ReturnAddress[00]:       0000000f3bf3f004 (pvz-portable + 0x16b004)
        ReturnAddress[01]:       0000000f3bf2d674 (pvz-portable + 0x159674)
        ReturnAddress[02]:       0000000f3bf8a344 (pvz-portable + 0x1b6344)
        ReturnAddress[03]:       0000000f3bf34618 (pvz-portable + 0x160618)
        ReturnAddress[04]:       0000000f3be92e0c (pvz-portable + 0xbee0c)
        ReturnAddress[05]:       0000000f3bdd8860 (pvz-portable + 0x4860)
        ReturnAddress[06]:       0000000f3bdd7700 (pvz-portable + 0x3700)
        ReturnAddress[07]:       0000000000000000
    Stack Dump:                               00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f
                                 00004a447cf0 60 7d 44 4a 00 00 00 00 04 f0 f3 3b 0f 00 00 00
                                 00004a447d00 e0 06 ff 6d 0c 00 00 00 e0 06 ff 6d 0c 00 00 00
                                 00004a447d10 00 80 f2 3b 0f 00 00 00 00 80 f2 3b 0f 00 00 00
                                 00004a447d20 10 90 e0 43 00 00 00 00 ff ff ff ff ff ff ff ff
                                 00004a447d30 26 00 00 00 3d 00 00 00 09 02 00 00 0c 00 00 00
                                 00004a447d40 58 81 65 3c 0f 00 00 00 80 81 65 3c 0f 00 00 00
                                 00004a447d50 20 82 65 3c 0f 00 00 00 00 00 00 00 00 00 00 00
                                 00004a447d60 00 7e 44 4a 00 00 00 00 74 d6 f2 3b 0f 00 00 00
                                 00004a447d70 e0 7b a1 65 0c 00 00 00 e0 06 ff 6d 0c 00 00 00
                                 00004a447d80 00 80 f2 3b 0f 00 00 00 00 80 f2 3b 0f 00 00 00
                                 00004a447d90 f0 7e 44 4a 00 00 00 00 00 00 00 00 00 00 00 00
                                 00004a447da0 10 90 e0 43 00 00 00 00 ff ff ff ff ff ff ff ff
                                 00004a447db0 f0 7e 44 4a 00 00 00 00 00 00 00 00 00 00 00 00
                                 00004a447dc0 10 90 e0 43 00 00 00 00 ff ff ff ff ff ff ff ff
                                 00004a447dd0 c0 7e 44 4a 00 00 00 00 00 80 f2 3b 0f 00 00 00
                                 00004a447de0 10 01 00 00 00 00 00 00 70 90 ab 6e 0c 00 00 00
    TLS Address:                 0000000028095200
    TLS Dump:                                 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f
                                 000028095200 00 00 01 00 09 08 e0 03 00 00 28 00 f0 79 44 4a
                                 000028095210 53 46 43 4f 00 00 00 00 00 00 00 00 99 5d e2 a4
                                 000028095220 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095230 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00
                                 000028095240 03 00 75 01 06 41 28 c0 00 00 00 00 00 00 00 00
                                 000028095250 28 00 75 01 f0 79 44 4a 00 00 28 00 00 00 00 00
                                 000028095260 00 00 00 00 18 00 00 00 58 7b 44 4a 00 00 18 00
                                 000028095270 f0 00 00 00 00 00 00 00 d0 00 00 00 00 00 00 00
                                 000028095280 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095290 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952a0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952b0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952c0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952d0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952e0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280952f0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Threads[01]:
    Thread ID:                   0000000000000b68
    Stack Region:                0000000067fe1000-0000000068002000
    Registers:
        X[00]:                   0000000c659fe1f0
        X[01]:                   0000000c659fe1f4
        X[02]:                   00000000007d81e2
        X[03]:                   ffffffffffffffff
        X[04]:                   0000000c659fe1f0
        X[05]:                   0000000000000000
        X[06]:                   0000000000000000
        X[07]:                   0000000000000000
        X[08]:                   0000000000000000
        X[09]:                   0000000000000000
        X[10]:                   0000000000000000
        X[11]:                   0000000000000000
        X[12]:                   0000000000000000
        X[13]:                   0000000000000000
        X[14]:                   0000000000000000
        X[15]:                   0000000000000000
        X[16]:                   0000000000000000
        X[17]:                   0000000000000000
        X[18]:                   0000000000000000
        X[19]:                   0000000000000000
        X[20]:                   0000000000000000
        X[21]:                   0000000000000000
        X[22]:                   0000000000000000
        X[23]:                   0000000000000000
        X[24]:                   0000000000000000
        X[25]:                   0000000000000000
        X[26]:                   0000000000000000
        X[27]:                   0000000000000000
        X[28]:                   0000000000000000
        FP:                      0000000068000e50
        LR:                      0000000f3c444dac (pvz-portable + 0x670dac)
        SP:                      0000000068000e50
        PC:                      0000000f3c43da94 (pvz-portable + 0x669a94)
    Stack Trace:
        ReturnAddress[00]:       0000000f3c4421cc (pvz-portable + 0x66e1cc)
        ReturnAddress[01]:       0000000f3c6419d0 (pvz-portable + 0x86d9d0)
        ReturnAddress[02]:       0000000f3c004c50 (pvz-portable + 0x230c50)
        ReturnAddress[03]:       0000000f3c001984 (pvz-portable + 0x22d984)
        ReturnAddress[04]:       0000000f3c001454 (pvz-portable + 0x22d454)
        ReturnAddress[05]:       0000000f3c0049dc (pvz-portable + 0x2309dc)
        ReturnAddress[06]:       0000000f3c4422f4 (pvz-portable + 0x66e2f4)
        ReturnAddress[07]:       0000000f3c445778 (pvz-portable + 0x671778)
        ReturnAddress[08]:       0000000000000000
    Stack Dump:                               00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f
                                 000068000e50 80 0e 00 68 00 00 00 00 cc 21 44 3c 0f 00 00 00
                                 000068000e60 f0 e1 9f 65 0c 00 00 00 2c 2c 2c 2c 2c 2c 2c 2c
                                 000068000e70 2c 2c 2c 2c 2c 2c 2c 2c 2c 2c 2c 2c 2c 2c 2c 2c
                                 000068000e80 90 0e 00 68 00 00 00 00 d0 19 64 3c 0f 00 00 00
                                 000068000e90 c0 0e 00 68 00 00 00 00 50 4c 00 3c 0f 00 00 00
                                 000068000ea0 f0 e1 9f 65 0c 00 00 00 84 19 00 3c 0f 00 00 00
                                 000068000eb0 ff ff ff ff 00 00 00 00 a0 49 95 3c 0f 00 00 00
                                 000068000ec0 e0 0e 00 68 00 00 00 00 84 19 00 3c 0f 00 00 00
                                 000068000ed0 ff ff ff ff 00 00 00 00 c4 19 00 3c 0f 00 00 00
                                 000068000ee0 30 0f 00 68 00 00 00 00 54 14 00 3c 0f 00 00 00
                                 000068000ef0 c0 86 a1 65 0c 00 00 00 c0 17 00 3c 0f 00 00 00
                                 000068000f00 a0 49 95 3c 0f 00 00 00 c0 86 a1 65 0c 00 00 00
                                 000068000f10 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000068000f20 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000068000f30 60 0f 00 68 00 00 00 00 dc 49 00 3c 0f 00 00 00
                                 000068000f40 50 7c 44 4a 00 00 00 00 80 55 09 28 00 00 00 00
    TLS Address:                 0000000028095400
    TLS Dump:                                 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f
                                 000028095400 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095410 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095420 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095430 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095440 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095450 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095460 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095470 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095480 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 000028095490 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280954a0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280954b0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280954c0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280954d0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280954e0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
                                 0000280954f0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00


### Steps to Reproduce

Start game

### Expected Behavior

works

### Actual Behavior

crash

### Logs / Screenshots

_No response_

### Additional Context

_No response_

__w/PvZ-Portable/PvZ-Portable/src/SexyAppFramework/graphics/GLPlatform.h:73:37: error: 'eglGetProcAddress' was not declared in this scope; did you mean 'SDL_GL_GetProcAddress'?
   73 |         gladLoadGLES2((GLADloadfunc)eglGetProcAddress);
      |                                     ^~~~~~~~~~~~~~~~~
      |                                     SDL_GL_GetProcAddress
make[2]: *** [CMakeFiles/pvz-portable.dir/build.make:303: CMakeFiles/pvz-portable.dir/src/Lawn/System/PoolEffect.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:112: CMakeFiles/pvz-portable.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
Error: Process completed with exit code 2.

> __w/PvZ-Portable/PvZ-Portable/src/SexyAppFramework/graphics/GLPlatform.h:73:37: error: 'eglGetProcAddress' was not declared in this scope; did you mean 'SDL_GL_GetProcAddress'? 73 | gladLoadGLES2((GLADloadfunc)eglGetProcAddress); | ^~~~~~~~~~~~~~~~~ | SDL_GL_GetProcAddress make[2]: *** [CMakeFiles/pvz-portable.dir/build.make:303: CMakeFiles/pvz-portable.dir/src/Lawn/System/PoolEffect.o] Error 1 make[1]: *** [CMakeFiles/Makefile2:112: CMakeFiles/pvz-portable.dir/all] Error 2 make: *** [Makefile:136: all] Error 2 Error: Process completed with exit code 2.

Try again with the latest commit.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -280,7 +280,7 @@ if (WIN32)
 		endif()
 	endif()
 elseif (NINTENDO_SWITCH)
-	target_link_libraries(pvz-portable PRIVATE EGL glapi drm_nouveau)
+	target_link_libraries(pvz-portable PRIVATE GLESv2 EGL glapi drm_nouveau)
 
 	nx_generate_nacp(pvz-portable.nacp
 		NAME "PvZ Portable"

--- a/src/SexyAppFramework/Common.cpp
+++ b/src/SexyAppFramework/Common.cpp
@@ -240,7 +240,7 @@ std::string Sexy::GetFullPath(const std::string& theRelPath)
 std::string Sexy::GetPathFrom(const std::string& theRelPath, const std::string& theDir)
 {
 	std::filesystem::path relPath = PathFromU8(theRelPath);
-	if (relPath.is_absolute() || relPath.has_root_name())
+	if (IsPathRooted(theRelPath))
 		return PathToU8(relPath.lexically_normal());
 
 	std::filesystem::path baseDir;
@@ -255,6 +255,37 @@ std::string Sexy::GetPathFrom(const std::string& theRelPath, const std::string& 
 	}
 
 	return PathToU8((baseDir / relPath).lexically_normal());
+}
+
+bool Sexy::IsPathRooted(std::string_view thePath)
+{
+	if (thePath.empty())
+		return false;
+
+	const std::filesystem::path aPath = PathFromU8(std::string(thePath));
+	if (aPath.has_root_path())
+		return true;
+
+#if defined(__SWITCH__) || defined(__3DS__)
+	const size_t aColonPos = thePath.find(':');
+	if (aColonPos == std::string_view::npos || aColonPos == 0 || aColonPos + 1 >= thePath.size())
+		return false;
+
+	const char aSeparator = thePath[aColonPos + 1];
+	if (aSeparator != '/' && aSeparator != '\\')
+		return false;
+
+	for (size_t i = 0; i < aColonPos; i++)
+	{
+		const unsigned char aChar = static_cast<unsigned char>(thePath[i]);
+		if (!std::isalnum(aChar) && aChar != '_')
+			return false;
+	}
+
+	return true;
+#else
+	return false;
+#endif
 }
 
 bool Sexy::AllowAllAccess(const std::string& theFileName)

--- a/src/SexyAppFramework/Common.h
+++ b/src/SexyAppFramework/Common.h
@@ -33,6 +33,7 @@
 #include <cstdint>
 #include <ctime>
 #include <filesystem>
+#include <string_view>
 #include <type_traits>
 #include <bit>
 
@@ -161,6 +162,7 @@ std::string			RemoveTrailingSlash(const std::string& theDirectory);
 std::string			GetCurDir();
 std::string			GetFullPath(const std::string& theRelPath);
 std::string			GetPathFrom(const std::string& theRelPath, const std::string& theDir);
+bool				IsPathRooted(std::string_view thePath);
 bool				AllowAllAccess(const std::string& theFileName);
 
 // Read memory and then move the pointer

--- a/src/SexyAppFramework/graphics/GLInterface.cpp
+++ b/src/SexyAppFramework/graphics/GLInterface.cpp
@@ -25,9 +25,7 @@
 #define GLAD_GLES2_IMPLEMENTATION
 #include "graphics/GLPlatform.h"
 
-#ifndef NINTENDO_SWITCH
 #include <SDL.h>
-#endif
 
 #include "graphics/GLInterface.h"
 #include "graphics/GLImage.h"
@@ -1120,7 +1118,7 @@ void GLInterface::UpdateViewport()
 {
 	int vx = 0, vy = 0, vw, vh;
 
-#ifdef NINTENDO_SWITCH
+#ifdef __SWITCH__
 	int width = 1280, height = 720;
 #else
 	int width, height;
@@ -1235,7 +1233,7 @@ bool GLInterface::PreDraw()
 void GLInterface::Flush()
 {
 	gNumVertices = 0;
-#ifdef NINTENDO_SWITCH
+#ifdef __SWITCH__
 	eglSwapBuffers(mApp->mWindow, mApp->mSurface);
 #else
 	SDL_GL_SwapWindow((SDL_Window*)mApp->mWindow);

--- a/src/SexyAppFramework/graphics/GLPlatform.h
+++ b/src/SexyAppFramework/graphics/GLPlatform.h
@@ -25,6 +25,12 @@
 #ifndef __GLPLATFORM_H__
 #define __GLPLATFORM_H__
 
+#ifdef __SWITCH__
+#include <switch.h>
+#include <EGL/egl.h>
+#include <EGL/eglext.h>
+#endif
+
 #ifdef _MSC_VER
 #pragma warning(push)
 #pragma warning(disable: 4551) // glad generated code triggers this on MSVC
@@ -35,6 +41,8 @@
 #ifdef _MSC_VER
 #pragma warning(pop)
 #endif
+
+#include <SDL.h>
 
 // Shared macro definitions — identical keywords in GLSL ES 1.00 and GLSL 1.20
 #define GLSL_VERT_MACROS \
@@ -49,26 +57,13 @@
 // When true, a desktop GL compatibility context is in use and shaders
 extern bool gDesktopGLFallback;
 
-#ifdef NINTENDO_SWITCH
-
-#include <switch.h>
-#include <EGL/egl.h>
-#include <EGL/eglext.h>
-
 inline void PlatformGLInit()
 {
+#ifdef __SWITCH__
 	gladLoadGLES2((GLADloadfunc)eglGetProcAddress);
-}
-
 #else
-
-#include <SDL.h>
-
-inline void PlatformGLInit()
-{
 	gladLoadGLES2((GLADloadfunc)SDL_GL_GetProcAddress);
-}
-
 #endif
+}
 
 #endif // __GLPLATFORM_H__

--- a/src/SexyAppFramework/graphics/Graphics.cpp
+++ b/src/SexyAppFramework/graphics/Graphics.cpp
@@ -377,8 +377,8 @@ void Graphics::PolyFill(const Point *theVertexList, int theNumVertices, bool con
 
 	mDestImage->FillScanLines(aSpans, aSpanPos, mColor, mDrawMode);
 
-	delete ind;
-	delete mPFActiveEdgeList;
+	delete[] ind;
+	delete[] mPFActiveEdgeList;
 }
 
 void Graphics::PolyFillAA(const Point *theVertexList, int theNumVertices, bool convex)

--- a/src/SexyAppFramework/imagelib/ImageLib.cpp
+++ b/src/SexyAppFramework/imagelib/ImageLib.cpp
@@ -54,7 +54,7 @@ Image::Image()
 
 Image::~Image()
 {
-	delete mBits;
+	delete[] mBits;
 }
 
 int	Image::GetWidth()
@@ -404,7 +404,7 @@ Image* GetGIFImage(const std::string& theFileName)
 				if (comments == (char*)nullptr)
 					break;
 
-				delete comments;
+				delete[] comments;
 				break;
 			}
 			case 0xff:
@@ -536,7 +536,7 @@ Image* GetGIFImage(const std::string& theFileName)
 
 				colortable[i] = 0xFF000000 | (r << 16) | (g << 8) | (b);
 			}
-			delete colormap;
+			delete[] colormap;
 		}
 
 		/*if (opacity >= (int) colors)
@@ -804,10 +804,10 @@ Image* GetGIFImage(const std::string& theFileName)
 			if (QuantumTick(y,image->rows))
 			MagickMonitor(LoadImageText,y,image->rows);*/
 		}
-		delete pixel_stack;
-		delete suffix;
-		delete prefix;
-		delete packet;
+		delete[] pixel_stack;
+		delete[] suffix;
+		delete[] prefix;
+		delete[] packet;
 
 		delete[] colortable;
 

--- a/src/SexyAppFramework/imagelib/ImageLib.cpp
+++ b/src/SexyAppFramework/imagelib/ImageLib.cpp
@@ -1325,8 +1325,7 @@ static bool CheckSinglePath(std::string_view thePath)
 	}
 
 	const std::string aPathString(thePath);
-	const std::filesystem::path aFilePath = Sexy::PathFromU8(aPathString);
-	if (!aFilePath.has_root_path())
+	if (!Sexy::IsPathRooted(aPathString))
 	{
 		const auto& aResourceBase = Sexy::GetResourceFolder();
 		if (!aResourceBase.empty())

--- a/src/SexyAppFramework/paklib/PakInterface.cpp
+++ b/src/SexyAppFramework/paklib/PakInterface.cpp
@@ -52,14 +52,14 @@ std::string PakInterface::NormalizePakPath(std::string_view theFileName)
 {
 	std::filesystem::path aFilePath = Sexy::PathFromU8(std::string(theFileName));
 
-	// Make absolute paths relative to resource folder.
-	if (aFilePath.has_root_directory())
+	// Make rooted paths relative to resource folder.
+	if (Sexy::IsPathRooted(theFileName))
 	{
 		const std::string& aResourceFolder = Sexy::GetResourceFolder();
 		if (!aResourceFolder.empty())
 		{
 			std::filesystem::path aResPath = Sexy::PathFromU8(aResourceFolder);
-			auto [aResEnd, aFileIt] = std::mismatch(aResPath.begin(), aResPath.end(), 
+			auto [aResEnd, aFileIt] = std::mismatch(aResPath.begin(), aResPath.end(),
 			                                       aFilePath.begin(), aFilePath.end());
 			if (aResEnd == aResPath.end())
 			{
@@ -72,7 +72,7 @@ std::string PakInterface::NormalizePakPath(std::string_view theFileName)
 	}
 
 	std::string aResult = Sexy::PathToU8(aFilePath.lexically_normal());
-	
+
 	if (aResult.size() >= 2 && aResult[0] == '.' && aResult[1] == '/')
 		aResult = aResult.substr(2);
 	
@@ -205,7 +205,7 @@ PFILE* PakInterface::FOpen(const char* theFileName, const char* anAccess)
 
 	const std::string& aResourceBase = Sexy::GetResourceFolder();
 	FILE* aFP = nullptr;
-	if (!aResourceBase.empty() && !Sexy::PathFromU8(theFileName).has_root_directory())
+	if (!aResourceBase.empty() && !Sexy::IsPathRooted(theFileName))
 	{
 		aFP = fcaseopenat(aResourceBase.c_str(), theFileName, anAccess);
 	}

--- a/src/SexyAppFramework/platform/switch/Input.cpp
+++ b/src/SexyAppFramework/platform/switch/Input.cpp
@@ -57,7 +57,7 @@ void SexyAppBase::InitInput()
 
 bool SexyAppBase::StartTextInput(std::string& theInput)
 {
-	char buf[512];
+	char buf[512] = {0};
 
 	SwkbdConfig kbd;
 	swkbdCreate(&kbd, 0);
@@ -117,7 +117,7 @@ bool SexyAppBase::ProcessDeferredMessages(bool singleMessage)
 
 		for (auto& k : keyMaps)
 		{
-			if (kDown & k.first)
+			if (kUp & k.first)
 				mWidgetManager->KeyUp(k.second);
 		}
 	}

